scrapy/downloadermiddlewares/robotstxt.py

```diff
@@ -85,26 +88,25 @@ def _logerror(self, failure, request, spider):
                          extra={'spider': spider})
         return failure
 
-    def _parse_robots(self, response, netloc):
+    def _parse_robots(self, response, netloc, spider):
         self.crawler.stats.inc_value('robotstxt/response_count')
         self.crawler.stats.inc_value(
             'robotstxt/response_status_count/{}'.format(response.status))
-        rp = robotparser.RobotFileParser(response.url)
-        body = ''
-        if hasattr(response, 'text'):
-            body = response.text
-        else:  # last effort try
-            try:
-                body = response.body.decode('utf-8')
-            except UnicodeDecodeError:
-                # If we found garbage, disregard it:,
-                # but keep the lookup cached (in self._parsers)
-                # Running rp.parse() will set rp state from
-                # 'disallow all' to 'allow any'.
-                self.crawler.stats.inc_value('robotstxt/unicode_error_count')
-        # stdlib's robotparser expects native 'str' ;
-        # with unicode input, non-ASCII encoded bytes decoding fails in Python2
-        rp.parse(to_native_str(body).splitlines())
+        
+        try:
+            rp = load_object(self._parserimpl)(response.body, self.crawler)
+        except ImportError as e:
+            # For Python 2 compatibility
+            errmsg = e.msg if hasattr(e, 'msg') else e.message
+
+            missingmodule = re.match('No module named (.+)', errmsg).group(1).strip("'")
+            logger.warning('Unable to use %(robotparser)s . Do you have %(module)s installed?'
+                        'Falling back to the default robots.txt parser.',
+                        {'robotparser': self._parserimpl, 'module': missingmodule}, 
+                        exc_info=sys.exc_info(),
+                        extra={'spider': spider})
+            
+            rp = PythonRobotParser(response.body, self.crawler)
```

Comment:
Why do we need a fallback here? I think it user asks for a specific robots.txt parsing backend, and backend is not available, we should be raising an error, not a warning.